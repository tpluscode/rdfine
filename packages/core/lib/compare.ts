import { Term } from 'rdf-js'
import type { RdfResource } from '../RdfResource'

export function terms(left: Term, right: Term): boolean {
  return left && left.equals(right)
}

export function resources(compareGraph: boolean) {
  return (left: RdfResource, right: RdfResource): boolean => {
    if (!left || !right) {
      return false
    }

    const idsEqual = left.id.equals(right.id)
    if (!compareGraph) {
      return idsEqual
    }

    return idsEqual && left._graphId.equals(right._graphId)
  }
}
