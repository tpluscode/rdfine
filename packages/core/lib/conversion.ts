import { Literal } from 'rdf-js'
import rdf from '@rdf-esm/data-model'
import { xsd } from '@tpluscode/rdf-ns-builders'
import { GraphPointer } from 'clownface'

const trueLiteral: Literal = rdf.literal('true', xsd.boolean)

export function fromLiteral(type: BooleanConstructor | StringConstructor | NumberConstructor | DateConstructor, obj: GraphPointer) {
  if (type === Boolean) {
    return trueLiteral.equals(obj.term)
  }

  if (type === Number) {
    return Number.parseFloat(obj.value)
  }

  if (type === Date) {
    return new Date(Date.parse(obj.value))
  }

  return obj.value
}
