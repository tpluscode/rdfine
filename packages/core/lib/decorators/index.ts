import type { RdfResource } from '../../RdfResource'
import type { Constructor } from '../ResourceFactory'

// From the TC39 Decorators proposal
export interface ClassElement {
  kind: 'field'|'method'
  key: PropertyKey
  placement: 'static'|'prototype'|'own'
  initializer?: Function
  extras?: ClassElement[]
  finisher?: <T extends RdfResource>(clazz: Constructor<T>) => undefined | Constructor<T>
  descriptor?: PropertyDescriptor
}

// From the TC39 Decorators proposal
export interface ClassDescriptor {
  kind: 'class'
  elements: ClassElement[]
  finisher?: <T extends RdfResource>(clazz: Constructor<T>) => undefined | Constructor<T>
}

export { namespace } from './namespace'
export { property } from './property'
