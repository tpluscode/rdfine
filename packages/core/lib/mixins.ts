import { RdfResource } from '../RdfResource'
import { Constructor } from './ResourceFactory'

export function * mixins(resource: RdfResource): Generator<Constructor> {
  // eslint-disable-next-line @typescript-eslint/ban-types
  let ctor: Constructor | Function = resource.constructor

  while (ctor) {
    if ('__properties' in ctor) {
      yield ctor
    }

    ctor = Object.getPrototypeOf(ctor)
  }
}
