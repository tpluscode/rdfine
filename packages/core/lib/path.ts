import { NamedNode } from 'rdf-js'
import RDF from '@rdfjs/data-model'
import cf, { SingleContextClownface } from 'clownface'
import { NamespaceBuilder } from '@rdfjs/namespace'

export type PropRef = string | NamedNode
export type EdgeTraversalFactory = (ns: NamespaceBuilder) => EdgeTraversal

export interface EdgeTraversal {
  (subject: SingleContextClownface): SingleContextClownface[]
  predicate: NamedNode
  crossesGraphBoundaries: boolean
}

function namespacedPredicate(term: string, namespace?: NamespaceBuilder): NamedNode {
  if (!namespace) {
    throw new Error(`Cannot construct URI for property ${term}. Annotate the class with @namespace or use a NamedNode`)
  }

  return namespace[term]
}

function predicate(termOrString: PropRef, namespace?: NamespaceBuilder): NamedNode {
  if (typeof termOrString === 'string') {
    if (termOrString.match(/^(http|urn):\/\//)) {
      return RDF.namedNode(termOrString)
    }

    return namespacedPredicate(termOrString, namespace)
  } else {
    return termOrString
  }
}

function sameGraph(prop: NamedNode): EdgeTraversal {
  const edge: EdgeTraversal = (subject) => subject.out(prop).toArray()

  edge.predicate = prop
  edge.crossesGraphBoundaries = false

  return edge
}

function anyGraph(prop: NamedNode): EdgeTraversal {
  const edge: EdgeTraversal = subject => {
    const graphNodes = new Map<string, SingleContextClownface>()

    subject.out(prop).forEach(node => {
      const quadsWithSubject = subject.dataset.match(node.term)
      const quadsWithObject = subject.dataset.match(null, null, node.term)
      const allQuads = [...quadsWithSubject, ...quadsWithObject]

      allQuads.forEach((quad) => {
        if (!graphNodes.has(quad.graph.value)) {
          graphNodes.set(quad.graph.value, cf({
            dataset: subject.dataset,
            term: quad.subject,
            graph: quad.graph,
          }))
        }
      })
    })

    return [...graphNodes.values()]
  }

  edge.predicate = prop
  edge.crossesGraphBoundaries = true

  return edge
}

export function crossBoundaries(prop: PropRef): EdgeTraversalFactory {
  return ns => anyGraph(predicate(prop, ns))
}

export function toEdgeTraversals({ constructor }: any, path: (PropRef | EdgeTraversalFactory)[]): EdgeTraversal[] {
  const namespace = constructor.__ns

  return path.map(prop => {
    if (typeof prop === 'function') {
      return prop(namespace)
    }

    return sameGraph(predicate(prop, namespace))
  })
}
