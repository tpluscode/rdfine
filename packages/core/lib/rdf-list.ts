import { Clownface, SingleContextClownface } from 'clownface'
import { rdf } from '@tpluscode/rdf-ns-builders'
import { DatasetCore, Term } from 'rdf-js'
import { RdfResource } from '../RdfResource'

function isLast(node: Clownface): boolean {
  return rdf.nil.equals(node.term)
}

export function isList(node: Clownface): boolean {
  const isLastListNode = isLast(node)
  const isListNode = node.out([rdf.first, rdf.rest]).values.length === 2
  return isLastListNode || isListNode
}

export function enumerateList<T, D extends DatasetCore = DatasetCore>(parent: RdfResource<D>, listNode: SingleContextClownface<Term, D>, fromTerm: (obj: SingleContextClownface<Term, D>) => T | T[]): T[] {
  const items: T[] = []

  let current = listNode
  while (!isLast(current)) {
    const firstItem = current.out(rdf.first).map((first, index) => {
      if (index > 0) {
        throw new Error('Malformed RDF List with multiple rdf:first objects')
      }

      return fromTerm.call(parent, first)
    })

    firstItem.forEach(item => {
      if (Array.isArray(item)) {
        throw new Error('Malformed RDF List had multiple rdf:first objects')
      }

      items.push(item)
    })

    current = current.out(rdf.rest).map((quad, index) => {
      if (index > 0) {
        throw new Error('Malformed RDF List with multiple rdf:rest objects')
      }

      return quad
    })[0]
  }

  return items
}
