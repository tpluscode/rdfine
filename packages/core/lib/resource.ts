import { AnyPointer, GraphPointer } from 'clownface'
import { BlankNode, NamedNode } from '@rdfjs/types'
import type { Initializer, RdfResourceCore, ResourceIdentifier } from '../RdfResource'

export function getPointer(graph: AnyPointer, id?: ResourceIdentifier | string): GraphPointer<BlankNode | NamedNode> {
  if (id && typeof id !== 'string') {
    return graph.node(id)
  }

  if (!id || id.startsWith('_:')) {
    return graph.blankNode()
  }

  return graph.namedNode(id)
}

export function fromInitializer(resource: RdfResourceCore, initializer: Initializer<RdfResourceCore>): RdfResourceCore {
  const term = getPointer(resource.pointer, initializer.id)
  return resource._create(term, [], { initializer })
}
