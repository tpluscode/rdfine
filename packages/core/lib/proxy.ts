import { DatasetCore, Literal, Term } from 'rdf-js'
import { SingleContextClownface } from 'clownface'
import type { RdfResource } from '../RdfResource'
import * as rdfList from './rdf-list'
import type { ResourceIndexer } from './ResourceFactory'

function nodeToValue(target: RdfResource) {
  const fromTerm = (obj: SingleContextClownface): any => {
    switch (obj.term.termType) {
      case 'BlankNode':
      case 'NamedNode':
        if (rdfList.isList(obj)) {
          return rdfList.enumerateList(target, obj, fromTerm)
        }

        return target._create(obj, [], { parent: target })
      default:
        return obj.term
    }
  }

  return fromTerm
}

export function createProxy<T extends RdfResource<D>, D extends DatasetCore = DatasetCore>(resource: T): T & ResourceIndexer {
  return new Proxy(resource, {
    get(target, property) {
      if (property in target) {
        return (target as any)[property.toString()]
      }

      const objects = target._getObjects(property.toString())

      if (objects.values.length === 0) {
        return undefined
      }

      const results = objects.map(nodeToValue(target))

      return results.length === 1 ? results[0] : results
    },

    set(target: T, property: string | number | symbol, value: T | Literal | Array<T | Literal | null> | null): boolean {
      if (property in target) {
        (target as any)[property.toString()] = value
        return true
      }

      if (typeof property === 'number') {
        return false
      }

      const values = Array.isArray(value) ? value : [value]
      const valueNodes = values
        .reduce((values, value) => {
          if (!value || typeof value !== 'object') {
            return values
          }

          if (value && 'termType' in value) {
            return [...values, value]
          }

          return [...values, value.id]
        }, [] as Term[])

      const predicate = target._selfGraph.namedNode(property.toString())
      target._selfGraph.deleteOut(predicate)

      if (values.length) {
        target._selfGraph.addOut(predicate, valueNodes)
      }

      return true
    },
  }) as any
}
