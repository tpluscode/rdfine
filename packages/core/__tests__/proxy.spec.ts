import cf, { GraphPointer } from 'clownface'
import $rdf from 'rdf-ext'
import { literal } from '@rdf-esm/data-model'
import DatasetExt from 'rdf-ext/lib/Dataset'
import { createProxy } from '../lib/proxy'
import RdfResourceImpl, { RdfResource } from '../RdfResource'
import { Literal, NamedNode } from 'rdf-js'
import { property, ResourceIndexer } from '../index'
import { ex } from './_helpers'

describe('proxy', () => {
  let node: GraphPointer<NamedNode, DatasetExt>

  beforeEach(() => {
    node = cf({ dataset: $rdf.dataset() })
      .namedNode(ex.Proxied)

    node.addOut(ex.Prop1, ex.Proxied2, p => {
      p.addOut(ex.Prop2, node.literal('foo', 'bar'))
    })

    node.addOut(ex.multi, ['a', 'z'])
  })

  describe('get', () => {
    it('proxies direct index access', () => {
      // given
      const resource = new RdfResourceImpl(node)
      const proxy = createProxy(resource)

      // when
      const value = proxy['http://example.com/Prop1'] as RdfResource

      // then
      expect(value.id.value).toEqual('http://example.com/Proxied2')
    })

    it('works with arbitrary symbol properties', () => {
      // given
      const Foo: unique symbol = Symbol('Foo')
      const proxy = createProxy(new RdfResourceImpl(node)) as any

      // when
      proxy[Foo] = 'bar'

      // then
      expect(proxy[Foo]).toEqual('bar')
    })

    it('returns raw literals', () => {
      // given
      const resource = new RdfResourceImpl(node)
      const proxy = createProxy(resource)

      // when
      const value = (proxy['http://example.com/Prop1'] as ResourceIndexer)['http://example.com/Prop2']

      // then
      expect(value).toEqual(node.literal('foo', 'bar').term)
    })

    it('returns array from multiple values', () => {
      // given
      const resource = new RdfResourceImpl(node)
      const proxy = createProxy(resource)

      // when
      const value = proxy['http://example.com/multi'] as Literal[]

      // then
      expect(value.map(l => l.value)).toStrictEqual(['a', 'z'])
    })

    it('does not proxy built-in properties', () => {
      // given
      const resource = new RdfResourceImpl(node)
      const proxy = createProxy(resource)

      // then
      expect(proxy.id.value).toEqual('http://example.com/Proxied')
    })

    it('does not proxy declared properties', () => {
      // given
      class Specialized extends RdfResourceImpl {
        @property.literal({ type: Number, path: 'http://example.com/number', initial: 5 })
        foo!: number
      }
      const proxy = createProxy(new Specialized(node))

      // then
      expect(proxy.foo).toStrictEqual(5)
    })

    it('returns undefined for missing property', () => {
      // given
      const resource = new RdfResourceImpl(node)
      const proxy = createProxy(resource)

      // when
      const value = proxy['http://example.com/foo']

      // then
      expect(value).toBeUndefined()
    })

    it('returns enumerated RDF list of literals', () => {
      // given
      const resource = new RdfResourceImpl(node)
      const proxy = createProxy(resource)
      node.addList(ex.listOfLiterals, ['a', 'B', 'cc'])

      // when
      const listItems = proxy['http://example.com/listOfLiterals'] as Literal[]

      // then
      expect(listItems.map(l => l.value)).toEqual(
        expect.arrayContaining(['a', 'B', 'cc']),
      )
    })

    it('returns enumerated RDF list of resources', () => {
      // given
      const resource = new RdfResourceImpl(node)
      const proxy = createProxy(resource)
      node.addList(ex.listOfUris, [
        ex.jane,
        ex.jane,
        ex.john,
      ])

      // when
      const listItems = proxy['http://example.com/listOfUris'] as RdfResource[]

      // then
      expect(listItems.map(l => l.id.value)).toEqual(
        expect.arrayContaining([ex.jane.value, ex.jane.value, ex.john.value]),
      )
    })

    it('attaches proxied resource as parent of objects', () => {
      // given
      const resource = new RdfResourceImpl(node)
      const proxy = createProxy(resource)
      node.addOut(ex.foo, ex.bar)

      // when
      const child = proxy[ex.foo.value] as RdfResource

      // then
      expect(child._parent).toBe(resource)
    })
  })

  describe('set', () => {
    it('single literal', () => {
      // given
      const resource = new RdfResourceImpl(node)
      const proxy = createProxy(resource)

      // when
      proxy[ex.set.value] = literal('foo')

      // then
      expect(node.dataset.toCanonical()).toMatchSnapshot()
    })

    it('single resource', () => {
      // given
      const resource = new RdfResourceImpl(node)
      const proxy = createProxy(resource)

      // when
      proxy[ex.set.value] = new RdfResourceImpl(node.namedNode(ex.other))

      // then
      expect(node.dataset.toCanonical()).toMatchSnapshot()
    })

    it('multiple literals', () => {
      // given
      const resource = new RdfResourceImpl(node)
      const proxy = createProxy(resource)

      // when
      proxy[ex.set.value] = [literal('foo'), literal('bar')]

      // then
      expect(node.dataset.toCanonical()).toMatchSnapshot()
    })

    it('multiple resources', () => {
      // given
      const resource = new RdfResourceImpl(node)
      const proxy = createProxy(resource)

      // when
      proxy[ex.set.value] = [
        new RdfResourceImpl(node.node(ex.one)),
        new RdfResourceImpl(node.node(ex.two)),
      ]

      // then
      expect(node.dataset.toCanonical()).toMatchSnapshot()
    })

    it('null removes triples', () => {
      // given
      const resource = new RdfResourceImpl(node)
      const proxy = createProxy(resource)

      // when
      proxy[ex.multi.value] = null

      // then
      expect(node.dataset.toCanonical()).toMatchSnapshot()
    })

    it('empty array removes triples', () => {
      // given
      const resource = new RdfResourceImpl(node)
      const proxy = createProxy(resource)

      // when
      proxy[ex.multi.value] = []

      // then
      expect(node.dataset.toCanonical()).toMatchSnapshot()
    })

    it('does not proxy declared properties', () => {
      // given
      class Specialized extends RdfResourceImpl {
        @property.literal({ type: Number, path: 'http://example.com/number', initial: 5 })
        foo!: number
      }
      const proxy = createProxy(new Specialized(node))

      // when
      proxy.foo = 10

      // then
      expect(node.dataset.toCanonical()).toMatchSnapshot()
    })
  })
})
