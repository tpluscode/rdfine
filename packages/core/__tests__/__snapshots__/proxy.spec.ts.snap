// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`proxy set does not proxy declared properties 1`] = `
"<http://example.com/Proxied2> <http://example.com/Prop2> \\"foo\\"@bar .
<http://example.com/Proxied> <http://example.com/Prop1> <http://example.com/Proxied2> .
<http://example.com/Proxied> <http://example.com/multi> \\"a\\" .
<http://example.com/Proxied> <http://example.com/multi> \\"z\\" .
<http://example.com/Proxied> <http://example.com/number> \\"10\\"^^<http://www.w3.org/2001/XMLSchema#integer> .
"
`;

exports[`proxy set empty array removes triples 1`] = `
"<http://example.com/Proxied2> <http://example.com/Prop2> \\"foo\\"@bar .
<http://example.com/Proxied> <http://example.com/Prop1> <http://example.com/Proxied2> .
"
`;

exports[`proxy set multiple literals 1`] = `
"<http://example.com/Proxied2> <http://example.com/Prop2> \\"foo\\"@bar .
<http://example.com/Proxied> <http://example.com/Prop1> <http://example.com/Proxied2> .
<http://example.com/Proxied> <http://example.com/multi> \\"a\\" .
<http://example.com/Proxied> <http://example.com/multi> \\"z\\" .
<http://example.com/Proxied> <http://example.com/set> \\"bar\\" .
<http://example.com/Proxied> <http://example.com/set> \\"foo\\" .
"
`;

exports[`proxy set multiple resources 1`] = `
"<http://example.com/Proxied2> <http://example.com/Prop2> \\"foo\\"@bar .
<http://example.com/Proxied> <http://example.com/Prop1> <http://example.com/Proxied2> .
<http://example.com/Proxied> <http://example.com/multi> \\"a\\" .
<http://example.com/Proxied> <http://example.com/multi> \\"z\\" .
<http://example.com/Proxied> <http://example.com/set> <http://example.com/one> .
<http://example.com/Proxied> <http://example.com/set> <http://example.com/two> .
"
`;

exports[`proxy set null removes triples 1`] = `
"<http://example.com/Proxied2> <http://example.com/Prop2> \\"foo\\"@bar .
<http://example.com/Proxied> <http://example.com/Prop1> <http://example.com/Proxied2> .
"
`;

exports[`proxy set single literal 1`] = `
"<http://example.com/Proxied2> <http://example.com/Prop2> \\"foo\\"@bar .
<http://example.com/Proxied> <http://example.com/Prop1> <http://example.com/Proxied2> .
<http://example.com/Proxied> <http://example.com/multi> \\"a\\" .
<http://example.com/Proxied> <http://example.com/multi> \\"z\\" .
<http://example.com/Proxied> <http://example.com/set> \\"foo\\" .
"
`;

exports[`proxy set single resource 1`] = `
"<http://example.com/Proxied2> <http://example.com/Prop2> \\"foo\\"@bar .
<http://example.com/Proxied> <http://example.com/Prop1> <http://example.com/Proxied2> .
<http://example.com/Proxied> <http://example.com/multi> \\"a\\" .
<http://example.com/Proxied> <http://example.com/multi> \\"z\\" .
<http://example.com/Proxied> <http://example.com/set> <http://example.com/other> .
"
`;
